pkg/host/utils.go

```diff
@@ -0,0 +1,184 @@
+	output, err := cmd.Output()
+	if err != nil {
+		log.Log.Error(err, "Failed to run mstvpd")
+		return "", "", err
+	}
+
+	// Parse the output for PN and SN
+	scanner := bufio.NewScanner(strings.NewReader(string(output)))
+	var partNumber, serialNumber string
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.HasPrefix(line, "PN:") {
+			partNumber = strings.TrimSpace(strings.TrimPrefix(line, "PN:"))
+		}
+		if strings.HasPrefix(line, "SN:") {
+			serialNumber = strings.TrimSpace(strings.TrimPrefix(line, "SN:"))
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		log.Log.Error(err, "Error reading mstvpd output")
+		return "", "", err
+	}
+
+	return partNumber, serialNumber, nil
+}
+
+func (h *hostUtils) GetFirmwareVersionAndPSID(pciAddr string) (string, string, error) {
+	log.Log.Info("HostUtils.GetFirmwareVersionAndPSID()", "pciAddr", pciAddr)
+	cmd := exec.Command("mstflint", "-d", pciAddr, "q")
+	output, err := cmd.Output()
+	if err != nil {
+		log.Log.Error(err, "Failed to run mstflint")
+		return "", "", err
+	}
+
+	// Parse the output for PN and SN
+	scanner := bufio.NewScanner(strings.NewReader(string(output)))
+	var firmwareVersion, PSID string
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.HasPrefix(line, "FW Version:") {
+			firmwareVersion = strings.TrimSpace(strings.TrimPrefix(line, "FW Version:"))
+		}
+		if strings.HasPrefix(line, "PSID:") {
+			PSID = strings.TrimSpace(strings.TrimPrefix(line, "PSID:"))
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		log.Log.Error(err, "Error reading mstflint output")
+		return "", "", err
+	}
+
+	return firmwareVersion, PSID, nil
+}
+
+func (h *hostUtils) GetLinkType(name string) string {
+	log.Log.Info("HostUtils.GetLinkType()", "name", name)
+	link, err := netlink.LinkByName(name)
+	if err != nil {
+		log.Log.Error(err, "GetLinkType(): failed to get link", "device", name)
+		return ""
+	}
+	return encapTypeToLinkType(link.Attrs().EncapType)
+}
+
+func encapTypeToLinkType(encapType string) string {
+	if encapType == "ether" {
+		return consts.ETHERNET
+	} else if encapType == "infiniband" {
+		return consts.INFINIBAND
+	}
+	return ""
+}
+
+func (h *hostUtils) GetInterfaceName(pciAddr string) string {
+	log.Log.Info("HostUtils.GetInterfaceName()", "pciAddr", pciAddr)
+
+	names, err := dputils.GetNetNames(pciAddr)
+	if err != nil || len(names) < 1 {
+		log.Log.Error(err, "GetInterfaceName(): failed to get interface name")
+		return ""
+	}
+	log.Log.Info("Interface name", "pciAddr", pciAddr, "name", names[0])
+	return names[0]
+}
+
+func (h *hostUtils) IsSriovVF(pciAddr string) bool {
+	log.Log.Info("HostUtils.IsSriovVF()", "pciAddr", pciAddr)
+
+	totalVfFilePath := filepath.Join(PciDevicesPath, pciAddr, "physfn")
+	if _, err := os.Stat(totalVfFilePath); err != nil {
+		return false
+	}
+	return true
+}
+
+func (h *hostUtils) GetRDMADeviceName(pciAddr string) string {
+	log.Log.Info("HostUtils.GetRDMADeviceName()", "pciAddr", pciAddr)
+
```

Comment:
can we leverage rdmamap lib for it ?

https://github.com/Mellanox/rdmamap/blob/d8230af2dc589a7e0a421e7b9927284d838d878a/rdma_map.go#L386

Reply:
Done